src/unix/fs.c

```diff
@@ -1148,8 +1148,19 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
 
   dst_flags = O_WRONLY | O_CREAT | O_TRUNC;
 
-  if (req->flags & UV_FS_COPYFILE_EXCL)
+  if (req->flags & UV_FS_COPYFILE_EXCL) {
     dst_flags |= O_EXCL;
+  } else {
+    /* If the source and the destination are the same, and the file is not being
+       opened exclusively, the file will be incorrectly truncated. Try to stat()
+       the destination. If it exists, and it is the same as the source, then
+       bail out early. */
+    if (stat(req->new_path, &dst_statsbuf) == 0 &&
```

Comment:
That's an interesting one. Could you, please, explain why previous logic wasn't correct?

Reply:
The previous logic was checking if the files were the same, but it was doing so after the destination had already been opened and truncated.